WikiWatershed/model-my-watershed
# Cache Model Run Results to Scenarios

## Overview

Enables reading saved results from scenario when requesting existing project or scenario from server, thus reducing requirement for repeated model runs.

Previously, loading a saved project incurred extraneous model run requests, one per scenario:

![image](https://cloud.githubusercontent.com/assets/1430060/8629701/ae95236e-272b-11e5-8f6a-71fb847647d1.png)

Now we do not re-run the model if there are no changes:

![image](https://cloud.githubusercontent.com/assets/1430060/8629709/c06ffb54-272b-11e5-83fe-0f32b40ed090.png)

However, if for whatever reason a scenario does not have a saved result, then the model is run on project load, and the results subsequently saved:

![image](https://cloud.githubusercontent.com/assets/1430060/8629733/05008d1a-272c-11e5-8928-a73d1c8db98d.png)

The scenario is **always** saved after a model run, even if a model run failed. In this case, the results would be set to `[]`, but other changes, such as inputs and modifications, will be saved to the database. This allows the user to re-open the project at a later date, at which time the model will be re-run with the saved inputs and modifications from last time.

~~Also, by saving only after a model run, we minimize duplicate save requests for the same scenario, instead grouping all changes into one `PUT`.~~

Since model runs can potentially take a long time, we also save the scenario inputs and modifications before starting the model run. This takes care of the edge case when a user closes the browser window before the model run completes, and loses their work. Whenever there are changes to `inputs` or `modifications`, we update the corresponding `inputmod_hash` and `modification_hash` fields. When starting model runs, we pass in these hashes, and they are included in the model output, which is then saved to the `modeling_scenario.results` field. When loading an existing project, we check to see if the `results` field is empty, and if it isn't, if the `inputmod_hash` matches that on the scenario, indicating a _stale_ result. If the `results` are empty or stale, we run the model, else we do nothing since everything is up to date.

## Data Changes

We had so far not marked `modeling_scenario.results` as a `JsonField`. Thus the existing data was incompatible, since it had the DRF `u` prefixed strings that was fixed for the modifications field in #320. Thus, we have a [data migration](https://docs.djangoproject.com/en/1.8/topics/migrations/#data-migrations) that sets the `results` field to `[]` for each row that had the `u` prefix. Thus, when such a scenario is loaded the next time, its results will be recalculated and saved in the correct format for the future.

We also add the `modeling_scenario.inputmod_hash` field in a second migration.

## Testing Instructions

 1. Run bundle and migrations `./scripts/bundle.sh && ./scripts/manage.sh migrate`
 2. Log in as a user and create a new project.
 3. Reload this existing project. Ensure that model calculations are not run for freshly loaded / unchanged project.
 4. Make arbitrary changes to the scenario: add modifications, change the slider, jump around between tabs. Ensure that after you stop and refresh the page, the previous state is recreated faithfully.
 5. Load an old project (if you have any on your system) which would have had its `results` emptied as part of the migration. Ensure that model calculations are run and saved. Reload the same project and ensure that model calculations are not re-run.
 6. Log out, and create a new project. Ensure that while model calculations are run, the scenario is not saved. Make arbitrary changes and then log in to save the project. Reload the project and ensure that model calculations were not run.
 7. Log out, and open a public project. Make changes, adjust the slider, ensure that the scenario is not saved after model calculations are run by reloading the page and finding it restored to its original state.
 8. Ensure that the model output graphs are rendered correctly when loading existing projects, and after model calculations are run.

Connects #548
